removing extra item on initial load in for blog index page

danielfmiranda opened the pull request:
Related PRs/issues: #9190
# Description

In this PR I introduced a if statement on the BlogIndexPage model, that will remove the extra item from the initial entry set. 

The if statement checks:
- If the page is filtered (meaning we wouldnt be on the landing page with the callout box, meaning no update is necessary)
- If the page size is set to 12 or 24 (the page sizes where this problem is occuring)

I only check for 12 or 24 because these are the page sizes that are divisible by 3, which is what the columns are set to. 4 and 8 always have extra items on the row, though the page size of 8 loads in nicely since the empty space is taken by the callout box. 

# Screenshots:


 **Screenshot of bug:**

<img width="300" alt="Screen Shot 2022-08-08 at 1 09 22 PM" src="https://user-images.githubusercontent.com/18314510/183799641-b3326a69-6c9f-4dff-9674-00a72854a79f.png">



**Page Size 4:**

Note that the initial entry count is unaffected, as page size of 4 or 8 are going to have overflowing cards anyway.
<img width="500" alt="Screen Shot 2022-08-08 at 1 09 22 PM" src="https://user-images.githubusercontent.com/18314510/183799023-cecf5186-2791-4079-91d0-dd40220e55bb.png">

<img width="500" alt="Screen Shot 2022-08-08 at 1 09 22 PM" src="https://user-images.githubusercontent.com/18314510/183799026-744f8ed7-efdb-4580-b2e7-eccc9d5f34ad.png">

**Page Size 12:**

Note that the initial entry count is - 1 so there is no extra card, and then loading more cards renders 4 rows as expected
<img width="500" alt="Screen Shot 2022-08-08 at 1 09 22 PM" src="https://user-images.githubusercontent.com/18314510/183799047-afcc1ef5-6c36-4882-abd2-58b1849153c2.png">
<img width="500" alt="Screen Shot 2022-08-08 at 1 09 22 PM" src="https://user-images.githubusercontent.com/18314510/183799050-63dccfee-b68c-40da-9d5a-73c3a05c2686.png">




# Checklist

<!-- Check off items with `[x]` or cross out items that don't apply with `~~The description~~` -->

**Tests**
- [ ] ~~Is the code I'm adding covered by tests?~~

**Changes in Models:**
- [ ] ~~Did I update or add new fake data?~~
- [ ] ~~Did I squash my migration?~~
- [ ] ~~Are my changes [backward-compatible](https://github.com/mozilla/foundation.mozilla.org/blob/main/docs/workflow.md#django-migrations-what-to-do-when-working-on-backward-incompatible-migrations). If not, did I schedule a deploy with the rest of the team?~~

**Documentation:**
- [ ] ~~Is my code documented?~~
- [ ] ~~Did I update the READMEs or wagtail documentation?~~

danielfmiranda commented on network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py:
```diff
@@ -155,6 +155,13 @@ class BlogIndexPage(IndexPage):
         context = super().get_context(request)
         context["related_topics"] = self.get_related_topics()
+
+        if not hasattr(self, 'filtered') and self.page_size == 12 or self.page_size == 24:
```
Making sure we are not on a filtered page, which do not render the callout box at all.

Also making sure that the page size is 12/24, which are the only page sizes divisible by 3 and having this issue.

tbrlpld replied:
It would be good to move this check into a helper method that clearly states what this check is supposed to mean. Something like `is_showing_callout_box_in_entries`. 

Also, we probably don't want to show the callout box if there are not topics.

danielfmiranda replied:
Great Idea! Updated and also added the check for the topics

danielfmiranda commented on network-api/networkapi/wagtailpages/pagemodels/index.py:
```diff
@@ -202,6 +203,9 @@ def generate_entries_set_html(self, request, *args, **kwargs):
 
         start = page * page_size
+        if 'entries_offset_for_callout' in context:
+            start = start - 1
```
If the BlogIndexPages initial entries were offset by the callout box (for example setting initial entry list size from 12 -> 11), we then need to update the starting point of the next page.

If we dont, the next page is going to start at 13, leaving entry #12 out of the rendered list.

tbrlpld replied:
**Question:** Do we use callout boxes in other index pages than the blog index? If not, we should not make this change on the index page, but rather keep it to the blog index.

tbrlpld replied:
Moving this to the blog index would also mean you don't need to pull in the context that you don't really need here.

tbrlpld replied:
I think it would be good to add some tests around this to cover all the edge cases, but I think in general this approach (of adjusting the start point) is fine.

danielfmiranda commented on network-api/networkapi/wagtailpages/templates/wagtailpages/blog_index_page.html:
```diff
@@ -75,7 +75,7 @@ <h2 class="tw-hidden tw-uppercase tw-text-3xl tw-mb-0 tw-font-bold tw-font-sans
       {% include "./fragments/entry_cards_item_loop.html" %}   
 
-      {% if not filtered %}
+      {% if not filtered and related_topics %}
```
If there are no topics set, do not render the topics box

danielfmiranda commented on network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py:
```diff
@@ -306,6 +313,68 @@ def extract_topic_information(self, topic_slug):
+        return related_topics
+
+    # Helper function to show if entry list is offset by topic box
+    def is_showing_topics_box_in_entry_list(self):
```
New helper function as suggested from feedback

danielfmiranda commented on network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py:
```diff
@@ -306,6 +313,68 @@ def extract_topic_information(self, topic_slug):
             'topic': topic_object
         }
 
+    def get_related_topics(self):
```
Moved this topic up here because I thought it would make sense to have all the routes at the end of the file.

danielfmiranda commented on network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py:
```diff
@@ -306,6 +313,68 @@ def extract_topic_information(self, topic_slug):
+        else:
+            return False
+
+    @route('^entries/')
```
Per request from feedback, I have moved the function for rendering new entries here, and simplified a little bit by removing any unneeded logic + adding the check for whether or not we need to update the index by 1.

danielfmiranda commented:
Thanks for the great feedback @tbrlpld!

Per request I have:
- Added some tests
- Created a helper function for checking whether or not the list is offset
- Updated blog index to load its new entry pages, instead of relying on index.py's

fessehaye reviewed (approved):
tests look good to me
